File: java/src/com/algos/topics/li10_tree/bt/revised/Bt_L1448_CountGoodNodesInBinaryTree.java
package  com.algos.topics.li10_tree.bt.revised;

import com.algos.topics.models.TreeNode;


public class Bt_L1448_CountGoodNodesInBinaryTree {

	public static void main(String[] args) {
		Bt_L1448_CountGoodNodesInBinaryTree obj = new Bt_L1448_CountGoodNodesInBinaryTree();

		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

		System.out.println(obj.goodNodes(root));
	}

    // Problem: https://leetcode.com/problems/count-good-nodes-in-binary-tree/
    // Idea: You need to track and pass the max value in each path recursive calls in top down.
	public int goodNodes(TreeNode root) {
        return goodNodesRec(root, Integer.MIN_VALUE);
    }

    private int goodNodesRec(TreeNode root, int max){
        if(root == null)
            return 0;

        int count = 0;
        if(root.val >= max){
            count = 1;
            max = root.val;
        }

        return count + goodNodesRec(root.left, max) + goodNodesRec(root.right, max);
    }

}

File: dsa/patterns/10_trees/bt/python/bt_l1448__count_good_nodes_in_binary_tree.py
# Binary Tree + DFS
# Problem: https://leetcode.com/problems/count-good-nodes-in-binary-tree/
# Idea: Track the max value along each path in recursive top-down DFS.
# Time: O(n)
# Space: O(h) recursion stack

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def good_nodes(root):
    def dfs(node, max_val):
        if not node:
            return 0
        count = 0
        if node.val >= max_val:
            count = 1
            max_val = node.val
        return count + dfs(node.left, max_val) + dfs(node.right, max_val)

    return dfs(root, float('-inf'))

# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(good_nodes(root))  # Expected: 4
